### PR TITLE
Treat Go raw string literals like ordinary string literals

### PR DESCRIPTION
**Pull request by nmote**
This addresses #3938, though this solution is a bit hacky so future work should be done to change the grammar so that this isn't necessary. I've posted more details in that issue.

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

**Comment by nmote** [comment]
I have two questions for reviewers:
1. Assuming this is acceptable at all, is there some utility I should use for manipulating the underlying string in a token, rather than pulling it apart myself?
2. It looks to me like the interpreted string literal code includes the locations of delimiters in the resulting location for the AST node, so am I correct in assuming that it's okay not to manipulate the location for the raw string literal here so that it omits the locations of the delimiters? I hope so, because making manual changes to the location would be very error-prone.

File: semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml
Comment (emjin) [inline semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:60]: Probably better to code defensively and add the check, since this depends on behavior of an external library

**Review by emjin — APPROVED** [review]
Thank you for taking this issue! This looks good to me! Leaving the tokens is great; that means autofix won't get messed up. Tagging @aryx or @IagoAbal to take a look in case I've missed some nicer way to do this.

**Comment by emjin** [comment]
Re the failing pre-commit check: https://semgrep.dev/docs/contributing/contributing-code/#development-workflow

File: semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml
Comment (nmote) [reply semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:60]: Fair enough, will do. I'll probably add a test that exercises an empty raw string and do a bounds-check with a `failwith`. That way any update that changes this behavior will get us a test failure. Sound good?

**Comment by nmote** [comment]
Thanks for the link; I missed that section. I'll get the formatting sorted out.

**Comment by nmote** [comment]
It looks like the pre-commit job needs approval from a maintainer. I did get it set up locally and it changed the formatting, so I think it should be green once it runs.

File: semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml
Comment (aryx) [inline semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml]: you can open Common and then do
if not (s =~ "^`\\(.*\\)`$")
then failwith "..."
else
  let str = Common.matched1 s in
  ...

Regexps are easier than those String.length, sub functions.

**Review by aryx — APPROVED** [review]
looks good! Thanks for fixing that.

File: semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml
Comment (IagoAbal) [inline semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:67]: We could simply add `EncodedString` here if we wanted to. The difference is that the way this is encoded, a pattern:
```
`foo`
```
will match a code expression `"foo"`, they are encoded as equivalent. For Python, Semgrep can match pattern `"foo"` against `r"foo"` via constant folding, but not the other way around.

I am not arguing in favor or against any approach, but I think we should be consistent.

Comment (nmote) [reply semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:67]: That makes sense, and it sounds like something that should be done. I'd prefer to leave it for another PR, though, if that's okay with you.

**Comment by nmote** [comment]
Thanks @aryx and @IagoAbal for the comments. I'll have time to make changes on Thursday.

File: semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml
Comment (IagoAbal) [reply semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:67]: I think the change is so small that it's better to do it in this PR, something like this should work:

```ocaml
Call(IdSpecial(EncodedString "r", tok), (tok,[str env tok],tok))
```

Otherwise soon enough somebody will complain about why this behaves "this way" in Python and "this other way" in Go.

Comment (nmote) [reply semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:67]: Okay, I will take a stab at that on Thursday. I only have an hour left on the time allotted for this project, though, so if I encounter any issues while making or testing this change it could put me over a bit.

Comment (IagoAbal) [reply semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:67]: Well, it may not be that small, because `string_literal` is not supposed to return an expression.

Comment (nmote) [reply semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:67]: I've looked into this and if I understand correctly, this would require a change to the pfff AST for Go. Right now, when we translate from the tree-sitter CST to the pfff AST, all string literals are represented by the `String` variant of the `literal` type defined in `semgrep-core/src/pfff/lang_go/parsing/ast_go.ml`. We'd need another variant there to differentiate between kinds of string literals, so that we could then choose `EncodedString` rather than `String` in `go_to_generic.ml` when we encounter a raw string literal.

Unfortunately, I don't currently have the time to make such a change. I hope that you can agree that as-is, this PR is still a step forward and is worth merging, but I understand if you'd rather wait until someone is able to make your requested changes.

**Comment by nmote** [comment]
I've made the changes requested by @emjin and @aryx, but unfortunately don't have the time to make the changes requested by @IagoAbal (details in the unresolved comment thread above). I still need a maintainer to approve some of the CI jobs, but I think they should all be green.

**Comment by aryx** [comment]
Ok I'll merge this. Sorry for our late reply, it was hack week last week so we got super busy.

**Review by aryx — APPROVED** [review]

File: semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml
Comment (IagoAbal) [reply semgrep-core/src/parsing/tree_sitter/Parse_go_tree_sitter.ml:67]: Yeah that sounds right! I think @aryx wants to merge anyways. No big deal for me. I just thought it's better to fix it consistently with current behavior for other languages, unless this is blocking some customer (which is not AFAIK). Good work anyways!

**Pull request merged**
